client/src/main/scala/org/http4s/client/middleware/DigestClient.scala

```diff
@@ -0,0 +1,180 @@
+    def digestAuthHeader(req: Request[F], challenge: Challenge): F[Header] =
+      for {
+        nonce <- F.fromOption(
+          challenge.params.get("nonce"),
+          new IllegalStateException(s"No nonce param in challenge, $challenge"))
+        (cnonce, ncLong) <- nonceCounter.nc(nonce)
+      } yield {
+        val realm = challenge.realm
+        val method = req.method.name
+        val uri = req.uri.path
+
+        val nc = ncLong.formatted("%08X")
+        val response =
+          DigestUtil.computeResponse(method, username, realm, password, uri, nonce, nc, cnonce, qop)
+        val params: NonEmptyList[(String, String)] = NonEmptyList.of(
+          "username" -> username,
+          "realm" -> realm,
+          "nonce" -> nonce,
+          "uri" -> uri,
+          "qop" -> qop,
+          "nc" -> nc,
+          "cnonce" -> cnonce,
+          "response" -> response,
+          "method" -> method
+        )
+        Authorization(Credentials.AuthParams(AuthScheme.Digest, params))
+      }
+
+    def prepareLoop(req: Request[F]): F[Resource[F, Response[F]]] =
+      F.continual(client.run(req).allocated) {
+        case Right((resp, dispose)) if !hasDigestAuth(req.headers) && resp.status == Unauthorized =>
+          extractChallenge(resp.headers) match {
+            case Some(challenge) =>
+              dispose >> digestAuthHeader(req, challenge) >>= (h => prepareLoop(req.putHeaders(h)))
+            case None =>
+              Resource.make(resp.pure[F])(_ => dispose).pure[F]
+          }
+        case Right((resp, dispose)) => Resource.make(resp.pure[F])(_ => dispose).pure[F]
+        case Left(e) => F.raiseError(e)
+      }
+
+    Client(req => Resource.suspend(prepareLoop(req)))
+  }
+
+  private def extractChallenge(headers: Headers): Option[Challenge] =
+    for {
+      challenge <- headers.get(`WWW-Authenticate`).map(_.value)
+      res <-
+        HttpHeaderParser
+          .WWW_AUTHENTICATE(challenge)
+          .fold(
+            _ => {
+              logger.info(s"Couldn't parse challenge: $challenge")
+              none
+            },
+            v => Some(v.values.head))
+    } yield res
+
+  private def hasDigestAuth(headers: Headers): Boolean =
+    headers
+      .get(Authorization)
+      .exists(_.credentials.authScheme == AuthScheme.Digest)
+
+}
+
+trait NonceCounter[F[_]] {
+
+  /** Returns `cnonce` and next `nc` value for the given nonce */
+  def nc(nonce: String): F[(String, Long)]
+
+  /** Call when a conversation is over, to prevent unbounded growth. */
+  def evict(nonce: String): F[Unit]
+
+  /** Call when a conversation is over, to prevent unbounded growth. */
+  def evictAll: F[Unit]
+}
+
+object NonceCounter {
+  val Max: Long = java.lang.Long.parseLong("FFFFFFFF", 16)
+  def nextCnonce(): String = {
+    val bytes = new Array[Byte](16)
+    Random.nextBytes(bytes)
+    Base64.getEncoder.encodeToString(bytes)
+  }
+  def defaultNonceCounter[F[_]: Concurrent: Timer]: NonceCounter[F] =
```

Comment:
`nextCnonce` is impure, which is a reminder that this counter is also impure.  We need `F[NonceCounter[F]]` for referential transparency, which ripples up to the `apply` methods returning `F[Client[F]]`.  This is a bit more annoying to construct, but also a bit more honest, and dishonesty has a way of eventually biting us in FP.